pkg/imgutil/dockerconfigresolver/credentialsstore.go

```diff
@@ -0,0 +1,181 @@
+	}
+
+	return nil, nil
+}
+
+// Store will save credentials for a given registry
+// On error, ErrUnableToStore
+func (cs *CredentialsStore) Store(registryURL *RegistryURL, credentials *Credentials) error {
+	// We just overwrite the server property here with the host
+	// Whether it was one of the variants, or was not set at all (see for example Amazon ECR, https://github.com/containerd/nerdctl/issues/733
+	// - which is likely a bug in docker) it doesn't matter.
+	// This is the credentials that were returned for that host, by the docker credentials store.
+	if registryURL.Namespace != nil {
+		credentials.ServerAddress = fmt.Sprintf("%s%s?%s", registryURL.Host, registryURL.Path, registryURL.RawQuery)
+	} else {
+		credentials.ServerAddress = registryURL.Host
+	}
+
+	// XXX future namespaced url likely require special handling here
+	if err := cs.dockerConfigFile.GetCredentialsStore(registryURL.CanonicalIdentifier()).Store(*(credentials)); err != nil {
+		return errors.Join(ErrUnableToStore, err)
+	}
+
+	return nil
+}
+
+// ShellCompletion will return candidate strings for nerdctl logout
+func (cs *CredentialsStore) ShellCompletion() []string {
+	candidates := []string{}
+	for key := range cs.dockerConfigFile.AuthConfigs {
+		candidates = append(candidates, key)
+	}
+
+	return candidates
+}
+
+// FileStorageLocation will return the file where credentials are stored for a given registry, or the empty string
+// if it is stored / to be stored in a different place (like an OS keychain, with docker credential helpers)
+func (cs *CredentialsStore) FileStorageLocation(registryURL *RegistryURL) string {
+	if store, isFile := (cs.dockerConfigFile.GetCredentialsStore(registryURL.CanonicalIdentifier())).(isFileStore); isFile {
+		return store.GetFilename()
+	}
+
+	return ""
+}
+
+// Retrieve gets existing credentials from the store for a certain registry.
+// If none are found, an empty Credentials struct is returned.
+// If we hard-fail reading from the store, indicative of a broken system, we wrap the error with ErrUnableToRetrieve
+func (cs *CredentialsStore) Retrieve(registryURL *RegistryURL, checkCredStore bool) (*Credentials, error) {
+	var err error
+	returnedCredentials := &Credentials{}
+
+	// As long as we depend on .ServerAddress, make sure it is populated correctly
+	// It does not matter what was stored - the docker cli clearly has issues with this
+	// What matters is that the credentials retrieved from the docker credentials store are *for that registryURL*
+	// and that is what ServerAddress should point to
+	defer func() {
+		if registryURL.Namespace != nil {
+			returnedCredentials.ServerAddress = fmt.Sprintf("%s%s?%s", registryURL.Host, registryURL.Path, registryURL.RawQuery)
+		} else {
+			returnedCredentials.ServerAddress = registryURL.Host
+		}
+	}()
+
+	if !checkCredStore {
+		return returnedCredentials, nil
+	}
+
+	log.L.Debugf("retrieving credentials from store for %s", registryURL.CanonicalIdentifier())
```

Comment:
This is all fairly generous with debug statements.

LMK if you want a lot of that out or if they should stay.